onnx_tf/handlers/backend/dilated_maxpooling.py

```diff
@@ -0,0 +1,430 @@
+        # implement it using: a % b = a - (a // b) * b
+
+        # inRow = (ind // num_channels) // output_width
+        # inCol = (ind // num_channels) % output_width
+        # ind_channel = ind % num_channels
+
+        ind_ = ind // num_channels
+        inRow = ind_ // output_width
+        inCol = ind_ - (ind_ // output_width) * output_width
+
+        ind_channel = ind - ind_ * num_channels
+
+        row = self._calc_input_ind(inRow, self.kernel_shape[0],
+                                   self.dilations[0],
+                                   self.strides[0]) - self.pads[0]
+        col = self._calc_input_ind(inCol, self.kernel_shape[1],
+                                   self.dilations[1],
+                                   self.strides[1]) - self.pads[2]
+
+        new_ind = num_channels * (row * in_width + col) + ind_channel
+        return new_ind
+
+    def _tf_product(self, a, b):
+        """
+            Calculates the cartesian product of two vectors a and b
+        """
+        tile_b = tf.tile(tf.expand_dims(b, 1), [1, tf.shape(a)[0]])
+        tile_b = tf.expand_dims(tile_b, 2)
+        tile_b = tf.reshape(tile_b, [-1, 1])
+
+        a = tf.tile(a, [tf.shape(b)[0], 1])
+        a = tf.concat([tile_b, a], axis=1)
+
+        return a
+
+    def _reduce_dilations(self):
+        """
+            This method reduces the dilations by extracting the values from
+            the input for every sliding window according to the dilations,
+            strides and kernel size and generates output that can be used by
+            pooling operation with strides = kernel_shape to accomplish
+            dilated pooling
+
+            Example:
+              Input:     [[  0,  1,  2,  3],
+                          [  4,  5,  6,  7],
+                          [  8,  9, 10, 11],
+                          [ 12, 13, 14, 15]]
+
+              Kernel:    [2, 2]
+              Dilations: [2, 2]
+              Strides:   [1, 1]
+
+              Will return:
+                         [[  0,  2,  1,  3],
+                          [  8, 10,  9, 11],
+                          [  4,  6,  5,  7],
+                          [ 12, 14, 13, 15]]
+
+              After max_pool2d with kernel_shape = strides = [2, 2]
+              the result is:
+                         [[ 10, 11],
+                          [ 14, 15]]
+
+              The method will also pad the input according to the paddings
+              provided and if ceil_mode is enabled
+        """
+
+        input_shape = _tf_shape(self.input)
+        in_spatial_shape = input_shape[1:self.spatial_size+1]
+
+        channel_num = input_shape[self.spatial_size+1]
+        gather_ind = tf.range(channel_num, dtype=tf.int64)
+        gather_ind = tf.expand_dims(gather_ind, 1)
+
+        self.output_shape = np.zeros([self.spatial_size + 2])
+        self.output_shape[0] = input_shape[0]
+        for dim in range(self.spatial_size - 1, -1, -1):
```

Comment:
A bit more inline comments in this function (_reduce_dilations) will be greatly appreciated.
I'm having a hard time understanding the detailed implementation here.

Reply:
OK, second look I got what you are doing here; but it would still be nice to explain that you are building a list of N-D index to extract values to reduce-pool. Also please add in comment some high-level design here, like you are calculating the output index and transform it back to input index by reusing the _calc_input_ind function. As it's a nice and simple trick worth explaining.

Reply:
Added comments in multiple places and tried to make the code mode readable.